extension/chrome/elements/compose-modules/compose-send-btn-module.ts

```diff
@@ -194,65 +214,62 @@ export class ComposeSendBtnModule extends ViewModule<ComposeView> {
     return { mimeType, data };
   };
 
-
-  private doSendMsg = async (msg: SendableMsg) => {
+  private doSendMsgs = async (msgObj: MultipleMessages): Promise<SendMsgsResult> => {
     // if this is a password-encrypted message, then we've already shown progress for uploading to backend
     // and this requests represents second half of uploadable effort. Else this represents all (no previous heavy requests)
     const progressRepresents = this.view.pwdOrPubkeyContainerModule.isVisible() ? 'SECOND-HALF' : 'EVERYTHING';
-    let msgSentRes: GmailRes.GmailMsgSend;
-    try {
-      this.isSendMessageInProgress = true;
-      msgSentRes = await this.view.emailProvider.msgSend(msg, (p) => this.renderUploadProgress(p, progressRepresents));
-    } catch (e) {
-      if (msg.thread && ApiErr.isNotFound(e) && this.view.threadId) { // cannot send msg because threadId not found - eg user since deleted it
-        msg.thread = undefined;
-        msgSentRes = await this.view.emailProvider.msgSend(msg, (p) => this.renderUploadProgress(p, progressRepresents));
-      } else {
-        this.isSendMessageInProgress = false;
-        throw e;
-      }
-    }
-    BrowserMsg.send.notificationShow(this.view.parentTabId, { notification: `Your ${this.view.isReplyBox ? 'reply' : 'message'} has been sent.` });
-    BrowserMsg.send.focusBody(this.view.parentTabId); // Bring focus back to body so Gmails shortcuts will work
-    const operations = [this.view.draftModule.draftDelete()];
-    if (msg.externalId) {
-      operations.push((async (externalId, id) => {
-        const gmailMsg = await this.view.emailProvider.msgGet(id, 'metadata');
-        const messageId = GmailParser.findHeader(gmailMsg, 'message-id');
-        if (messageId) {
-          await this.view.acctServer.messageGatewayUpdate(externalId, messageId);
-        } else {
-          Catch.report('Failed to extract Message-ID of sent message');
+    const sentIds: string[] = [];
+    const supplementaryOperations: Promise<void>[] = [];
+    const supplementaryOperationsErrors: any[] = []; // tslint:disable-line:no-unsafe-any
+    const success: EmailParts[] = [];
+    const failures: { recipient: EmailParts, e: any }[] = [];
+    for (const msg of msgObj.msgs) {
+      let msgSentRes: GmailRes.GmailMsgSend;
+      const msgRecipients = msg.getAllRecipients();
+      while (true) { // first try with msg.thread, and then possibly try again without it
```

Comment:
Could we use some different approach here than `while(true)` ?  It functions correctly here but doesn't seem like what the `while` statement was meant for.

To be honest, I'm not entirely sure what the different way could be. Maybe a bit of recursion. The contents of `try { } catch { }` could be its own async method, and the catch could include a conditional call to itself. I figured you wanted to avoid that because you'd have to pass many arrays to it for count-keeping, which you can pass as one object as follows:

```ts
const sendMsgsResults: SendMsgsResult = {
  sentIds: [],
  supplementaryOperations: [],
  supplementaryOperationsErrors: [],
  success: [],
  failures: []
}  
```

Which could be passed into method, passed internally during recursion, then directly returned below.

I think that will make the code more readable.